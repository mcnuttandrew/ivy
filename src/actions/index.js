/* global chrome*/
const SECOND = 1000;
const MINUTE = SECOND * 60;
const HOUR = MINUTE * 60;
const DAY = HOUR * 24;

const buildEasyAction = type => payload => dispatch =>
  dispatch({type, payload});
export const clearSelection = buildEasyAction('clear-selection');

// example action
// export const setPageId = payload => dispatch => {
//   dispatch({type: 'set-page-id', payload});
//   getTreeForId(payload)
//     .then(result => dispatch({
//       type: 'get-tree-from-cache',
//       payload: {
//         data: result,
//         pageId: payload
//       }
//     }));
// };
