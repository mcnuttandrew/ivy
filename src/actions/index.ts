import {Dispatch} from 'redux';

import VegaDataPreAlias from 'vega-datasets';
const VegaData: {[key: string]: any} = VegaDataPreAlias;

import {Analyzer} from 'type-analyzer';
const {computeColMeta} = Analyzer;

// const SECOND = 1000;
// const MINUTE = SECOND * 60;
// const HOUR = MINUTE * 60;
// const DAY = HOUR * 24;

export interface GenericAction {
  (payload?: any): (dispatch: Dispatch) => void;
}
interface GenericActionCreator {
  (type: string): GenericAction;
}

const buildEasyAction: GenericActionCreator = type => payload => dispatch =>
  dispatch({type, payload});
// export const loadDataFromPredefinedDatasets = buildEasyAction(
//   'load-data-from-predefined',
// );

export const loadDataFromPredefinedDatasets: GenericAction = fileName => dispatch => {
  fetch(VegaData[fileName].url)
    .then(d => d.json())
    .then(d => {
      dispatch({
        type: 'recieve-data-from-predefined',
        payload: d,
      });

      dispatch({
        type: 'recieve-type-inferences',
        payload: computeColMeta(d),
      });
    });
};

export const changeSelectedFile: GenericAction = fileName => dispatch => {
  dispatch({
    type: 'change-selected-file',
    payload: fileName,
  });
  loadDataFromPredefinedDatasets(fileName)(dispatch);
};

// example action
// export const setPageId = payload => dispatch => {
//   dispatch({type: 'set-page-id', payload});
//   getTreeForId(payload)
//     .then(result => dispatch({
//       type: 'get-tree-from-cache',
//       payload: {
//         data: result,
//         pageId: payload
//       }
//     }));
// };
