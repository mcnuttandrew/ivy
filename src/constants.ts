// i know this file is bad and crazy i feel bad
export interface OptionValue {
  display: string;
  value: string;
}
export interface EncodingOption {
  optionType: string;
  options: OptionValue[];
  optionSetter: (spec: any, selectedOption: string) => any;
  optionGetter: (spec: any) => string;
  optionDefault: string;
  predicate: (spec: any) => boolean;
}

import {VegaMark} from './types';
export const marks: VegaMark[] = [
  'arc',
  'area',
  // 'image',
  // 'group',
  // 'line',
  // 'path',
  'rect',
  // 'rule',
  // 'shape',
  'symbol',
  'text',
  // 'trail',
];

// check if using the nested spec mode, if so, re-route into that subspec
const maybePrefixWithSpec = (spec: any, route: string[]): string[] =>
  Boolean(spec.getIn(['spec'])) ? ['spec', ...route] : route;

const justCountAgg = [
  {display: 'none', value: undefined},
  {display: 'count', value: 'count'},
];
const spatialAggs = [
  ...justCountAgg,
  {display: 'min', value: 'min'},
  {display: 'max', value: 'max'},
  {display: 'sum', value: 'sum'},
  {display: 'mean', value: 'mean'},
  {display: 'median', value: 'median'},
  {display: 'mode', value: 'mode'},
];
const toOption = (x: string): OptionValue => ({display: x, value: x});
const binningOptions = [
  'none',
  'count',
  'min',
  'mean',
  'sum',
  'bin', // TODO bin is handled incorrectly
  'max',
  'median',
].map(toOption);

const buildSpatialOptions = (
  dimension: string,
  options: OptionValue[],
): EncodingOption => ({
  optionType: 'Agg.',
  options,
  optionSetter: (spec, selectedOption) => {
    const route = maybePrefixWithSpec(spec, ['encoding', dimension]);
    const newSpec = spec.setIn([...route, 'aggregate'], selectedOption);
    const addType = (x: any) => x.setIn([...route, 'type'], 'quantitative');
    const hasField = newSpec.getIn([...route, 'field']);
    return hasField ? newSpec : addType(newSpec);
  },
  optionGetter: spec => {
    const route = maybePrefixWithSpec(spec, [
      'encoding',
      dimension,
      'aggregate',
    ]);
    return spec.getIn(route);
  },
  optionDefault: 'none',
  predicate: () => true,
});

const typeRoute = (dim: string) => ['encoding', dim, 'scale', 'type'];
const buildScaleOption = (dim: string): EncodingOption => ({
  optionType: 'Scale type',
  options: ['linear', 'log'].map(toOption),
  optionSetter: (spec, option) =>
    spec.setIn(maybePrefixWithSpec(spec, typeRoute(dim)), option),
  optionGetter: spec =>
    spec.getIn(maybePrefixWithSpec(spec, typeRoute(dim))) || 'linear',
  optionDefault: 'linear',
  predicate: spec =>
    Boolean(spec.getIn(maybePrefixWithSpec(spec, ['encoding', dim, 'field']))),
});

const buildCoerceRoute = (spec: any, dim: string): string[] =>
  maybePrefixWithSpec(spec, ['encoding', dim, 'type']);
const buildTypeCoercion = (dim: string): EncodingOption => ({
  optionType: 'Data type',
  options: ['nominal', 'ordinal', 'quantitative', 'temporal'].map(toOption),
  optionSetter: (spec, option) =>
    spec.setIn(buildCoerceRoute(spec, dim), option),
  optionGetter: spec => spec.getIn(buildCoerceRoute(spec, dim)),
  optionDefault: undefined,
  predicate: () => true,
});

// take in an encoding option and create a predicate that rejects that option
// if that field is not present
type predicateInject = (dim: string, option: EncodingOption) => EncodingOption;
const buildPredRoute = (spec: any, dim: string): string[] =>
  maybePrefixWithSpec(spec, ['encoding', dim, 'field']);
const injectFieldPred: predicateInject = (dim, option) => ({
  ...option,
  predicate: (spec: any): boolean => !!spec.getIn(buildPredRoute(spec, dim)),
});
const injectNofieldPred: predicateInject = (dim, option) => ({
  ...option,
  predicate: (spec: any): boolean => !spec.getIn(buildPredRoute(spec, dim)),
});

const generateXorY = (dim: string) => [
  injectFieldPred(dim, buildTypeCoercion(dim)),
  buildScaleOption(dim),
  injectFieldPred(dim, buildSpatialOptions(dim, spatialAggs)),
  injectNofieldPred(dim, buildSpatialOptions(dim, justCountAgg)),
];

// the configuration options consumed by the encoding panel
export const configurationOptions: any = {
  x: generateXorY('x'),
  y: generateXorY('y'),
  'row-meta': [],
  'column-meta': [],
  'repeat-meta': [],

  // repeaters
  facet: [injectFieldPred('facet', buildTypeCoercion('facet'))],
  row: [injectFieldPred('row', buildTypeCoercion('row'))],
  column: [injectFieldPred('column', buildTypeCoercion('column'))],
  // marks
  size: [
    injectFieldPred('size', buildTypeCoercion('size')),
    injectFieldPred('size', buildSpatialOptions('size', binningOptions)),
    injectNofieldPred('size', buildSpatialOptions('size', justCountAgg)),
  ],
  color: [
    injectFieldPred('color', buildTypeCoercion('color')),
    injectFieldPred('color', buildSpatialOptions('color', binningOptions)),
    injectNofieldPred('color', buildSpatialOptions('color', justCountAgg)),
  ],
  shape: [
    injectFieldPred('shape', buildTypeCoercion('shape')),
    injectFieldPred('shape', buildSpatialOptions('shape', binningOptions)),
  ],
  detail: [
    injectFieldPred('detail', buildTypeCoercion('detail')),
    injectFieldPred('detail', buildSpatialOptions('detail', binningOptions)),
  ],
  text: [
    injectFieldPred('text', buildTypeCoercion('text')),
    injectFieldPred('text', buildSpatialOptions('text', binningOptions)),
    injectNofieldPred('text', buildSpatialOptions('text', justCountAgg)),
  ],
};
