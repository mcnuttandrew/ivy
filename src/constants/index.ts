export const test = 'test';
type ON = 'on';
export const EDITOR_OPTIONS = {
  selectOnLineNumbers: true,
  automaticLayout: true,
  wordWrap: 'on' as ON,
  minimap: {
    enabled: false,
  },
  fontSize: 10,
  lineDecorationsWidth: 0,
  lineNumbersMinChars: 3,
};

export const TEXT_TYPE: {[x: string]: string} = {
  MEASURE: '123',
  DIMENSION: 'ABC',
  TIME: 'ðŸ•‘',
};

export const GRAMMAR_NAME = 'grammer';
export const GRAMMAR_DESC = 'Tableau-style grammar of graphics';

// export const
export const WIDGET_VALUES = 'Settings';
export const WIDGET_CONFIGURATION = 'Params';
export const JSON_OUTPUT = 'Output';
export const TEMPLATE_BODY = 'Body';

/* eslint-disable no-undef*/
export const PREVENT_ACCIDENTAL_LEAVE = process.env.NODE_ENV === 'production';
/* eslint-enable no-undef*/
