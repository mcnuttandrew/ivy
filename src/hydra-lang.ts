import {
  Template,
  TemplateMap,
  TemplateWidget,
  WidgetValidationQuery,
  DataTargetWidget,
  WidgetSubType,
  ListWidget,
  Shortcut,
  SwitchWidget,
  SliderWidget,
} from './templates/types';
import {trim} from './utils';

/**
 * Evaluate a hydra query, used for the widget validations and conditional checks
 * // TODO: this system doesn't support data type checking?
 * @param query - for now uses the special widget validation langugage
 * @param templateMap - the specification/variable values defined by the gui
 */
function evaluateQuery(query: WidgetValidationQuery, templateMap: TemplateMap): boolean {
  // TODO add a type check function to this
  const generatedContent = new Function('parameters', `return ${query}`);
  return Boolean(generatedContent(templateMap));
}

export function evaluateShortcut(shortcut: Shortcut, templateMap: TemplateMap): TemplateMap {
  const generatedContent = new Function('parameters', `return ${shortcut.shortcutFunction}`);
  const newMap = generatedContent(templateMap);
  return newMap;
}

interface ConditionalArgs {
  query: string;
  true: any;
  false: any;
  deleteKeyOnFalse?: boolean;
  deleteKeyOnTrue?: boolean;
}
function shouldUpdateContainerWithValue(
  queryResult: 'true' | 'false',
  conditional: ConditionalArgs,
): boolean {
  // if a conditional doesn't want that value to get added to the traversing object then ingnore it
  return (
    (queryResult === 'false' && conditional.deleteKeyOnFalse) ||
    (queryResult === 'true' && conditional.deleteKeyOnTrue)
  );
}

// syntax example
// {...example,
// mark: {
//   type: 'point',
//   tooltip: true,
//   color: {CONDITIONAL: {true: '[Single Color]', false: null, query: {Color: null}}},
// }}
/**
 * Walk across the tree and apply conditionals are appropriate,
 * example conditional syntax: {CONDITIONAL: {true: '[Single Color]', false: null, query: '!parameters.Color}}
 *
 * @param templateMap - the specification/variable values defined by the gui
 */
export function applyConditionals(templateMap: TemplateMap): any {
  return function walker(spec: any): any {
    // if it's array interate across it
    if (Array.isArray(spec)) {
      return spec.reduce((acc, child) => {
        if (child && typeof child === 'object' && child.CONDITIONAL) {
          const queryResult = evaluateQuery(child.CONDITIONAL.query, templateMap) ? 'true' : 'false';
          if (!shouldUpdateContainerWithValue(queryResult, child.CONDITIONAL)) {
            return acc.concat(walker(child.CONDITIONAL[queryResult]));
          } else {
            return acc;
          }
        }
        return acc.concat(walker(child));
      }, []);
    }
    // check if it's null or not an object return
    if (!(typeof spec === 'object' && spec !== null)) {
      return spec;
    }
    // if the object being consider is itself a conditional evaluate it
    if (typeof spec === 'object' && spec.CONDITIONAL) {
      const queryResult = evaluateQuery(spec.CONDITIONAL.query, templateMap) ? 'true' : 'false';
      if (!shouldUpdateContainerWithValue(queryResult, spec.CONDITIONAL)) {
        return walker(spec.CONDITIONAL[queryResult]);
      } else {
        return null;
      }
    }
    // otherwise looks through its children
    return Object.entries(spec).reduce((acc: any, [key, value]: any) => {
      // if it's a conditional, if so execute the conditional
      if (value && typeof value === 'object' && value.CONDITIONAL) {
        const queryResult = evaluateQuery(value.CONDITIONAL.query, templateMap) ? 'true' : 'false';
        if (!shouldUpdateContainerWithValue(queryResult, value.CONDITIONAL)) {
          acc[key] = walker(value.CONDITIONAL[queryResult]);
        }
      } else {
        acc[key] = walker(value);
      }
      return acc;
    }, {});
  };
}

/**
 *
 * @param template
 * @param templateMap - the specification/variable values defined by the gui
 */
export function applyQueries(template: Template, templateMap: TemplateMap): TemplateWidget<any>[] {
  const widgetMap = template.widgets.reduce((acc: any, widget) => {
    acc[widget.widgetName] = true;
    return acc;
  }, {});

  const validWidgetNames = template.widgetValidations.reduce((acc: {[x: string]: boolean}, validation) => {
    const queryResult = evaluateQuery(validation.query, templateMap);
    acc[validation.queryTarget] = validation.queryResult === 'show' ? queryResult : !queryResult;
    return acc;
  }, widgetMap);
  return template.widgets.filter(widget => validWidgetNames[widget.widgetName]);
}

/**
 * Apply values from templatemap (specification) to template
 * Important to do it this way and not via json parse because values might be parts of strings
 * Example {"field": "datum.[VARIABLENAME]"}
 * @param code
 * @param templateMap - the specification/variable values defined by the gui
 */
export const setTemplateValues = (code: string, templateMap: TemplateMap): string => {
  const filledInSpec = Object.entries(templateMap).reduce((acc: string, keyValue: any) => {
    const [key, value] = keyValue;
    if (trim(value) !== value) {
      // this supports the weird HACK required to make the interpolateion system
      // not make everything a string
      return acc
        .replace(new RegExp(`"\\[${key}\\]"`, 'g'), value || 'null')
        .replace(new RegExp(`\\[${key}\\]`, 'g'), trim(value) || 'null');
    }
    const reg = new RegExp(`"\\[${key}\\]"`, 'g');
    return acc.replace(reg, (Array.isArray(value) && JSON.stringify(value)) || value || 'null');
  }, code);
  return filledInSpec;
};

/**
 * Generate a list of the missing fields on a template
 * @param template
 * @param templateMap - the specification/variable values defined by the gui
 */
export function getMissingFields(template: Template, templateMap: TemplateMap): string[] {
  const requiredFields = template.widgets
    .filter(d => d.widgetType === 'DataTarget' && (d as TemplateWidget<DataTargetWidget>).widget.required)
    .map(d => d.widgetName);
  const missingFileds = requiredFields
    .map((fieldName: string) => ({fieldName, value: !templateMap[fieldName]}))
    .filter((d: any) => d.value)
    .map(d => d.fieldName);

  return missingFileds;
}

/**
 * Figure out if the all of the required fields have been filled in
 * @param template
 * @param templateMap - the specification/variable values defined by the gui
 */
export function checkIfMapComplete(template: Template, templateMap: TemplateMap): boolean {
  const missing = getMissingFields(template, templateMap);
  return missing.length === 0;
}

/**
 * for template map holes that are NOT data columns, fill em as best you can
 * @param template
 */
export function constructDefaultTemplateMap(template: Template): TemplateMap {
  return template.widgets.reduce((acc: any, w: TemplateWidget<WidgetSubType>) => {
    let value = null;
    if (w.widgetType === 'MultiDataTarget') {
      value = [];
    }
    if (w.widgetType === 'Text') {
      return acc;
    }
    if (w.widgetType === 'List') {
      value = (w as TemplateWidget<ListWidget>).widget.defaultValue;
    }
    if (w.widgetType === 'Switch') {
      const localW = w as TemplateWidget<SwitchWidget>;
      value = localW.widget.defaultsToActive ? localW.widget.activeValue : localW.widget.inactiveValue;
    }
    if (w.widgetType === 'Slider') {
      value = (w as TemplateWidget<SliderWidget>).widget.defaultValue;
    }
    acc[w.widgetName] = value;
    return acc;
  }, {});
}

/**
 *
 * @param template
 * @param templateMap - the specification/variable values defined by the gui
 */
export function evaluateHydraProgram(template: Template, templateMap: TemplateMap): any {
  // 1. apply variables to string representation of code
  const interpolatedVals = setTemplateValues(template.code, templateMap);
  // 2. parse to json
  // const parsedJson = JSON.parse(interpolatedVals);
  let parsedJson = null;
  try {
    parsedJson = JSON.parse(interpolatedVals);
  } catch (e) {
    console.log(e, 'crash');
    parsedJson = {};
  }
  // 3. evaluate inline conditionals
  const evaluatableSpec = applyConditionals(templateMap)(parsedJson);
  // 4. return
  return evaluatableSpec;
}
