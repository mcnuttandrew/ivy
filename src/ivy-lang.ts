import {
  ConditionQuery,
  DataTargetWidget,
  GenWidget,
  ListWidget,
  MultiDataTargetWidget,
  Shortcut,
  SliderWidget,
  SwitchWidget,
  Template,
  TemplateMap,
  Widget,
} from './types';
import {trim, get, log, logError} from './utils';
import {JsonMap, Json, JsonArray} from './types';

interface ConditionalArgs {
  query: string;
  true?: Json;
  false?: Json;
}
export type IvyLangConditional = {$cond: ConditionalArgs};

/**
 * Evaluate a ivy query, used for the widget conditions and conditional checks
 * // TODO: this system doesn't support data type checking?
 * @param query - for now uses the special widget condition langugage
 * @param templateMap - the specification/variable values defined by the gui
 */
function evaluateQuery(query: ConditionQuery, templateMap: TemplateMap): boolean {
  // TODO add a type check function to this
  // TODO can probable keep a cache of these results?
  let result = false;
  try {
    const generatedContent = new Function('parameters', `return ${query}`);
    result = Boolean(generatedContent(templateMap.paramValues));
  } catch (e) {
    log('Query Evalution Error', e, query, templateMap.paramValues);
  }
  return result;
}

export function evaluateShortcut(shortcut: Shortcut, templateMap: TemplateMap): TemplateMap {
  let newMap = templateMap;
  try {
    const generatedContent = new Function('parameters', `return ${shortcut.shortcutFunction}`);
    newMap = generatedContent(templateMap.paramValues);
  } catch (e) {
    log('Short cut error', e, shortcut.shortcutFunction, shortcut.label);
  }
  return newMap;
}

function shouldUpdateContainerWithValue(
  queryResult: 'true' | 'false',
  conditional: ConditionalArgs,
): boolean {
  // if a conditional doesn't want that value to get added to the traversing object then ingnore it
  return !Object.keys(conditional).includes(queryResult);
}

// syntax example
// {...example,
// mark: {
//   type: 'point',
//   tooltip: true,
//   color: {$cond: {true: '[Single Color]', false: null, query: {Color: null}}},
// }}
// TODO: a lot of the logic in this could be cleaned up into a pretty combinator pattern
/**
 * Walk across the tree and apply conditionals are appropriate,
 * example conditional syntax: {$cond: {true: '[Single Color]', false: null, query: '!parameters.Color}}
 *
 * @param templateMap - the specification/variable values defined by the gui
 * @returns JSON (any is the dumb stand in for json)
 */
export function applyConditionals(templateMap: TemplateMap): (spec: Json) => Json {
  return function walker(spec: Json): Json {
    // if it's array interate across it
    if (Array.isArray(spec)) {
      return spec.reduce((acc: JsonArray, child) => {
        if (child && typeof child === 'object' && (child as JsonMap).$cond) {
          const valuemap = (child as unknown) as IvyLangConditional;
          const queryResult = evaluateQuery(valuemap.$cond.query, templateMap) ? 'true' : 'false';
          if (!shouldUpdateContainerWithValue(queryResult, valuemap.$cond)) {
            return acc.concat(walker(valuemap.$cond[queryResult]));
          } else {
            return acc;
          }
        }
        return acc.concat(walker(child));
      }, []);
    }
    // check if it's null or not an object return
    if (!(typeof spec === 'object' && spec !== null)) {
      return spec;
    }
    // if the object being consider is itself a conditional evaluate it
    if (typeof spec === 'object' && spec.$cond) {
      const valuemap = (spec as unknown) as IvyLangConditional;
      const queryResult = evaluateQuery(valuemap.$cond.query, templateMap) ? 'true' : 'false';
      if (!shouldUpdateContainerWithValue(queryResult, valuemap.$cond)) {
        return walker(valuemap.$cond[queryResult]);
      } else {
        return null;
      }
    }
    // otherwise looks through its children
    return Object.entries(spec).reduce((acc: JsonMap, [key, value]: [string, Json]) => {
      // if it's a conditional, if so execute the conditional
      if (value && typeof value === 'object' && (value as JsonMap).$cond) {
        const valuemap = (value as unknown) as IvyLangConditional;
        const queryResult = evaluateQuery(valuemap.$cond.query, templateMap) ? 'true' : 'false';
        if (!shouldUpdateContainerWithValue(queryResult, valuemap.$cond)) {
          acc[key] = walker(valuemap.$cond[queryResult]);
        }
      } else {
        acc[key] = walker(value);
      }
      return acc;
    }, {});
  };
}

/**
 *
 * @param template
 * @param templateMap - the specification/variable values defined by the gui
 */
export function applyQueries(template: Template, templateMap: TemplateMap): Widget<any>[] {
  return template.widgets.filter(widget => {
    if (!widget.conditions || !widget.conditions.length) {
      return true;
    }
    return widget.conditions.every(condition => {
      const queryResult = evaluateQuery(condition.query, templateMap);
      return condition.queryResult === 'show' ? queryResult : !queryResult;
    });
  });
}

/**
 * Apply values from templatemap (specification) to template
 * Important to do it this way and not via json parse because values might be parts of strings
 * Example {"field": "datum.[VARIABLENAME]"}
 * @param code
 * @param templateMap - the specification/variable values defined by the gui
 */
export const setTemplateValues = (code: string, templateMap: TemplateMap): string => {
  const filledInSpec = Object.entries(templateMap.paramValues).reduce(
    (acc: string, [key, value]: [string, string | null]) => {
      const reWith = new RegExp(`"\\[${key}\\]"`, 'g');
      const reWithout = new RegExp(`\\[${key}\\]`, 'g');
      if (trim(value) !== value) {
        // this supports the weird HACK required to make the interpolateion system
        // not make everything a string
        return acc.replace(reWith, value || 'null').replace(reWithout, trim(value) || 'null');
      }
      const setVal = (Array.isArray(value) && JSON.stringify(value)) || value || 'null';
      return acc.replace(reWith, setVal).replace(reWithout, setVal);
    },
    code,
  );
  return filledInSpec;
};

/**
 * Generate a list of the missing fields on a template
 * @param template
 * @param templateMap - the specification/variable values defined by the gui
 */
export function getMissingFields(template: Template, templateMap: TemplateMap): string[] {
  const params = templateMap.paramValues;

  // data target
  const missingFileds = template.widgets
    .filter(d => d.type === 'DataTarget' && (d as Widget<DataTargetWidget>).config.required)
    .filter(d => !params[d.name])
    .map(d => d.name);

  // multi data target
  const missingMultiFileds = template.widgets
    .filter(d => d.type === 'MultiDataTarget' && (d as Widget<MultiDataTargetWidget>).config.required)
    .filter(
      (widget: Widget<MultiDataTargetWidget>) =>
        !params[widget.name] || params[widget.name].length < widget.config.minNumberOfTargets,
    )
    .map(d => d.name);
  return missingFileds.concat(missingMultiFileds);
}

/**
 * Figure out if the all of the required fields have been filled in
 * @param template
 * @param templateMap - the specification/variable values defined by the gui
 */
export function checkIfMapComplete(template: Template, templateMap: TemplateMap): boolean {
  const missing = getMissingFields(template, templateMap);
  return missing.length === 0;
}

export function getDefaultValueForWidget(widget: GenWidget): any {
  if (widget.type === 'MultiDataTarget') {
    return [];
  }
  if (widget.type === 'Text' || widget.type === 'Section') {
    return null;
  }
  if (widget.type === 'List') {
    const config = (widget as Widget<ListWidget>).config;
    if (config.defaultValue) {
      return config.defaultValue;
    }
    const firstValue = config.allowedValues[0];
    if (!firstValue) {
      return null;
    }
    return typeof firstValue === 'string' ? firstValue : firstValue.value;
  }
  if (widget.type === 'Switch') {
    const localW = widget as Widget<SwitchWidget>;
    return localW.config.defaultsToActive ? localW.config.active : localW.config.inactive;
  }
  if (widget.type === 'Slider') {
    return (widget as Widget<SliderWidget>).config.defaultValue;
  }
  if (widget.type === 'FreeText') {
    return '';
  }
  return null;
}
const MARK_UP_COMPONENTS = new Set(['Text', 'Section']);

/**
 * for template map holes that are NOT data columns, fill em as best you can
 * @param template
 */
export function constructDefaultTemplateMap(template: Template): TemplateMap {
  const paramValues = template.widgets.reduce((acc: any, widget: GenWidget) => {
    if (MARK_UP_COMPONENTS.has(widget.type)) {
      return acc;
    }
    acc[widget.name] = getDefaultValueForWidget(widget);
    return acc;
  }, {});

  return {paramValues, systemValues: {viewsToMaterialize: {}, dataTransforms: []}};
}

/**
 *
 * @param template
 * @param templateMap - the specification/variable values defined by the gui
 */
export function evaluateIvyProgram(template: Template, templateMap: TemplateMap): Json {
  // 1. apply variables to string representation of code
  const interpolatedVals = setTemplateValues(template.code, templateMap);
  // 2. parse to json
  let parsedJson = null;
  try {
    parsedJson = JSON.parse(interpolatedVals);
  } catch (e) {
    logError('crash', e);
    logError('crashed on ', interpolatedVals);
    parsedJson = {};
  }
  // 3. evaluate inline conditionals
  const evaluatableSpec = applyConditionals(templateMap)(parsedJson);
  // 4. return
  return evaluatableSpec;
}

const DUMMY = 'xxxxxEXAMPLExxxx';
export function generateFullTemplateMap(widgets: GenWidget[]): TemplateMap {
  const paramValues = widgets.reduce((acc: {[x: string]: any}, widget: GenWidget) => {
    const widgetType = widget.type;
    if (widgetType === 'DataTarget') {
      acc[widget.name] = `"${DUMMY}"`;
    }
    if (widgetType === 'MultiDataTarget') {
      acc[widget.name] = `[${DUMMY}, ${DUMMY}]`;
    }
    if (widgetType === 'List') {
      const localW = widget as Widget<ListWidget>;
      acc[widget.name] = localW.config.defaultValue || get(localW, ['config', 'allowedValues', 0, 'value']);
    }
    if (widgetType === 'Switch') {
      const localW = widget as Widget<SwitchWidget>;
      acc[widget.name] = localW.config.active;
    }
    if (widgetType === 'Slider') {
      const localW = widget as Widget<SliderWidget>;
      acc[widget.name] = localW.config.defaultValue;
    }
    return acc;
  }, {});

  return {
    paramValues,
    systemValues: {viewsToMaterialize: {}, dataTransforms: []},
  };
}

/////////////////
///////////////// The following section is VERY cursed, be weary
/////////////////

const buildConditionalValidation = (type: any, defKey: string): any => {
  return {
    type: 'object',
    additionalProperties: false,
    properties: {
      $cond: {
        description:
          'In conditional arguments, requires a query ( which is a js predicate), a true branch, and a false branch',
        type: 'object',
        properties: {
          query: {
            type: 'string',
            description:
              'Javascript predicate to be evaluated to determine result of predicate, must return a boolean. Some tips:\n\n 1. The current settings can be accessed by using the parameters object\n 2. The values held by values in the parameters object are strings.\n\n\n For example if you wanted to checked if boolean parameter setting was currently set to true, a reasonable way to do that would be:\n\n `parameters.switchVal === "true"` \n\n NOT `parameters.switchVal`\n\n The latter will always evaluated to true (because strings in JS are truthy).',
          },
          true: type ? {$ref: `#/${defKey}/${type}`} : {type: 'string'},
          false: type ? {$ref: `#/${defKey}/${type}`} : {type: 'string'},
        },
        required: ['query'],
      },
    },
  };
};
const buildInterpolantType = (old: any): any => {
  const escapeType = {type: 'string', pattern: '\\[.*\\]'};
  if (old.anyOf) {
    return {anyOf: [escapeType, ...old.anyOf]};
  } else if (old.type) {
    return {anyOf: [escapeType, {...old}]};
  } else {
    return escapeType;
  }
};

export function modifyJSONSchema(jsonSchema: any): any {
  const defKey = jsonSchema.$schema === 'http://json-schema.org/draft-06/schema#' ? 'defs' : 'definitions';
  const conditionalItem = {$ref: `#/${defKey}/IvyConditional`};

  function schemaWalk(spec: any): any {
    if (!spec) {
      return spec;
    }
    // if it's array interate across void
    if (Array.isArray(spec)) {
      return spec.map(child => schemaWalk(child));
    }
    // check if it's null or not an object return
    if (typeof spec !== 'object') {
      return spec;
    }

    if (spec.type && spec.type !== 'object' && typeof spec.type !== 'object') {
      const nodeClone = JSON.parse(JSON.stringify(spec));
      delete nodeClone.type;
      delete nodeClone.enum;
      const typeCopy: any = {type: spec.type};
      if (spec.enum) {
        typeCopy.enum = spec.enum;
      }
      nodeClone.anyOf = [conditionalItem, typeCopy, buildInterpolantType(spec)];
      return nodeClone;
    }
    if (spec.anyOf) {
      const nodeClone = JSON.parse(JSON.stringify(spec));
      nodeClone.anyOf = nodeClone.anyOf
        .filter((d: any) => d.type !== 'null')
        .map((d: any) => buildInterpolantType(d))
        .concat(nodeClone.anyOf);
      return nodeClone;
    }

    // otherwise looks through its children
    return Object.entries(spec).reduce((acc: any, [key, value]: any) => {
      acc[key] = schemaWalk(value);
      return acc;
    }, {});
  }
  const newSchema = schemaWalk(jsonSchema);
  newSchema[defKey] = Object.entries(newSchema[defKey]).reduce((acc: any, [key, value]: any) => {
    acc[`${key}-core-props`] = value;
    acc[key] = {
      anyOf: [
        {$ref: `#/${defKey}/${key}-core-props`},
        buildConditionalValidation(`${key}-core-props`, defKey),
      ],
    };
    return acc;
  }, {});
  newSchema[jsonSchema.$ref] = jsonSchema[jsonSchema.$ref];

  newSchema[defKey].IvyConditional = buildConditionalValidation(false, defKey);
  return newSchema;
}
