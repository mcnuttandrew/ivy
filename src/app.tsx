import React from 'react';
import ReactDOM from 'react-dom';
import AppWrap from './components/app-wrap';
import setupMonaco from './utils/monaco';
import {PREVENT_ACCIDENTAL_LEAVE} from './constants/CONFIG';
import {retrieveThumbnails} from './thumbnail';

import './stylesheets/main.css';
import './stylesheets/rc-slider.css';
import 'rc-tooltip/assets/bootstrap_white.css';

setupMonaco();
retrieveThumbnails();

ReactDOM.render(<AppWrap />, document.querySelector('#root-container'));
if (document.querySelector('.loading-msg')) {
  document.querySelector('.loading-msg').remove();
}

if (PREVENT_ACCIDENTAL_LEAVE) {
  window.onbeforeunload = (): string => 'Are you sure you want to leave?';
}
