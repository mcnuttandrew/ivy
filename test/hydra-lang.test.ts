import {applyConditionals} from '../src/hydra-lang';
import Scatterplot from '../src/templates/example-templates/scatterplot';

test('#applyConditionals', () => {
  const PARSED_CODE = JSON.parse(Scatterplot.code);

  const exampleTemplateMap1: any = {Color: null};
  expect(applyConditionals(PARSED_CODE, exampleTemplateMap1)).toMatchSnapshot();

  const exampleTemplateMap2: any = {Color: 'Wowza good dimension'};
  expect(applyConditionals(PARSED_CODE, exampleTemplateMap2)).toMatchSnapshot();
});
