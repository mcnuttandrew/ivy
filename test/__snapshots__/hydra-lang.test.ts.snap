// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#applyConditionals 1`] = `
Object {
  "$schema": "https://vega.github.io/schema/vega-lite/v4.json",
  "encoding": Object {
    "color": Object {
      "field": "[Color]",
      "type": "[colorType]",
    },
    "x": Object {
      "field": "[xDim]",
      "scale": Object {
        "zero": "[Zeroes]",
      },
      "type": "[xType]",
    },
    "y": Object {
      "field": "[yDim]",
      "scale": Object {
        "zero": "[Zeroes]",
      },
      "type": "[yType]",
    },
  },
  "mark": Object {
    "color": "[Single Color]",
    "tooltip": true,
    "type": "point",
  },
}
`;

exports[`#applyConditionals 2`] = `
Object {
  "$schema": "https://vega.github.io/schema/vega-lite/v4.json",
  "encoding": Object {
    "color": Object {
      "field": "[Color]",
      "type": "[colorType]",
    },
    "x": Object {
      "field": "[xDim]",
      "scale": Object {
        "zero": "[Zeroes]",
      },
      "type": "[xType]",
    },
    "y": Object {
      "field": "[yDim]",
      "scale": Object {
        "zero": "[Zeroes]",
      },
      "type": "[yType]",
    },
  },
  "mark": Object {
    "color": null,
    "tooltip": true,
    "type": "point",
  },
}
`;
